test/e2e/dhcp_overrides_test.go

```diff
@@ -0,0 +1,124 @@
+
+import (
+	"encoding/base64"
+	"fmt"
+	"path/filepath"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	"github.com/vmware/govmomi/property"
+	"github.com/vmware/govmomi/vim25/mo"
+	"gopkg.in/yaml.v2"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/utils/pointer"
+	"sigs.k8s.io/cluster-api/test/framework/clusterctl"
+	"sigs.k8s.io/cluster-api/util"
+
+	types "sigs.k8s.io/cluster-api-provider-vsphere/apis/v1beta1"
+)
+
+type GuestInfoMetadata struct {
+	Network Network `yaml:"network"`
+}
+
+type Network struct {
+	Ethernets map[string]Ethernet `yaml:"ethernets"`
+}
+
+type Ethernet struct {
+	DHCP4Overrides DHCPOverrides `yaml:"dhcp4-overrides"`
+}
+
+type DHCPOverrides struct {
+	SendHostname *bool `yaml:"send-hostname"`
+}
+
+var _ = Describe("DHCPOverrides configuration test", func() {
+	When("Creating a cluster with DHCPOverrides configured", func() {
+		const specName = "dhcp-overrides"
+		var namespace *corev1.Namespace
+
+		BeforeEach(func() {
+			namespace = setupSpecNamespace(specName)
+		})
+
+		AfterEach(func() {
+			cleanupSpecNamespace(namespace)
+		})
+
+		It("Only configures the network with the provided nameservers", func() {
+			clusterName := fmt.Sprintf("%s-%s", specName, util.RandomString(6))
+			clusterResources := new(clusterctl.ApplyClusterTemplateAndWaitResult)
+
+			By("Creating a workload cluster")
+			clusterctl.ApplyClusterTemplateAndWait(ctx, clusterctl.ApplyClusterTemplateAndWaitInput{
+				ClusterProxy: bootstrapClusterProxy,
+				ConfigCluster: clusterctl.ConfigClusterInput{
+					LogFolder:                filepath.Join(artifactFolder, "clusters", bootstrapClusterProxy.GetName()),
+					ClusterctlConfigPath:     clusterctlConfigPath,
+					KubeconfigPath:           bootstrapClusterProxy.GetKubeconfigPath(),
+					InfrastructureProvider:   clusterctl.DefaultInfrastructureProvider,
+					Flavor:                   "dhcp-overrides",
+					Namespace:                namespace.Name,
+					ClusterName:              clusterName,
+					KubernetesVersion:        e2eConfig.GetVariable(KubernetesVersion),
+					ControlPlaneMachineCount: pointer.Int64Ptr(1),
+					WorkerMachineCount:       pointer.Int64Ptr(1),
+				},
+				WaitForClusterIntervals:      e2eConfig.GetIntervals(specName, "wait-cluster"),
+				WaitForControlPlaneIntervals: e2eConfig.GetIntervals(specName, "wait-control-plane"),
+				WaitForMachineDeployments:    e2eConfig.GetIntervals(specName, "wait-worker-nodes"),
+			}, clusterResources)
+
+			list := getVSphereVMsForCluster(clusterName, namespace.Name)
+			for _, vm := range list.Items {
+				metadata, err := getVMMetadata(vm)
+				Expect(err).NotTo(HaveOccurred())
+				guestInfoMetadata := &GuestInfoMetadata{}
+				err = yaml.Unmarshal(metadata, guestInfoMetadata)
+				Expect(err).NotTo(HaveOccurred())
+				Expect(guestInfoMetadata.Network.Ethernets["id0"].DHCP4Overrides.SendHostname).NotTo(BeNil())
+				Expect(*guestInfoMetadata.Network.Ethernets["id0"].DHCP4Overrides.SendHostname).To(BeFalse())
```

Comment:
From the `dhcp-overrides.yaml`, the override is being set only on the CP machine template. Update the kustomization to add it to both the templates, or a better test is to add different properties for the CP and worker node machine template.
When asserting them, the main difference would be that the CP VM would have the `cluster.x-k8s.io/control-plane: ""` label. So writing an assertion based off that would be the best thing to do.

Reply:
Thanks! Updated!